common/src/test/java/io/netty/util/ResourceLeakDetectorTest.java

```diff
@@ -98,6 +102,34 @@ private boolean closeResources(boolean checkClosed) {
         assertNoErrors(error);
     }
 
+    @Test
+    // triggering GC is too unreliable for regular testing.
+    @Disabled
+    public void testLeakSetupHints() throws Throwable {
+        leakResource();
+
+        // trigger GC
+        System.gc();
+        try {
+            @SuppressWarnings("unused")
+            byte[] unused = new byte[(int) Math.min(Runtime.getRuntime().totalMemory(), Integer.MAX_VALUE - 1000)];
```

Comment:
Ok I've just tried this but can't get this to work reliably either.

Relying on `GarbageCollectorMXBean` fails for two reasons: "scavenge" GC (young GC I assume?) has its own MX Bean, but might not collect all references. So that bean might show a young GC has happened, but references wouldn't be processed yet, because they're only done in old GC. I've also tried excluding beans with the name "scavenge", which made it more reliable, but still sometimes failed. I think this might be because references can be processed concurrently: https://developers.redhat.com/articles/2021/05/20/shenandoah-garbage-collection-openjdk-16-concurrent-reference-processing – not sure if this applies to non-shenandoah, but it looks like GC can sometimes do reference processing after the GC has already nominally completed, so we can't rely on the MX bean to tell us when references have been processed. It's also "fixed" by adding a `Thread.sleep(10)` just after the enqueue check, which corroborates this.

```java
    private static long getCollectionCount() {
        long count = 0;
        for (GarbageCollectorMXBean bean : garbageCollectorMxBeans) {
            if (bean.getName().equals("scavenge")) {
                continue;
            }
            count += bean.getCollectionCount();
        }
        return count;
    }

    private static void triggerGc() {
        Assumptions.assumeFalse(garbageCollectorMxBeans.isEmpty(), "GarbageCollectorMXBean not available");
        long collectionCountPrevious = getCollectionCount();

        System.gc();

        @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
        List<Object> storage = new ArrayList<Object>();
        try {
            while (getCollectionCount() == collectionCountPrevious) {
                storage.add(new byte[1024 * 1024]); // 1M
            }
        } catch (OutOfMemoryError ignored) {
        }
        storage.clear();

        try {
            Thread.sleep(10);
        } catch (InterruptedException e) {
        }
    }
```

Another approach I tried is to use a second `ReferenceQueue` with a reference that is definitely unreachable, and simply allocating until that reference has been enqueued. However even this approach was not 100% reliable (maybe ~80%?). I think this might also be due to concurrent reference processing, where there is a race condition between the GC thread enqueuing the reference, and the test thread checking for the reference's presence. The same "fix" of adding a 10ms delay works here too.

```java
    private static void triggerGc() {
        ReferenceQueue<Object> queue = new ReferenceQueue<Object>();
        Object obj = new Object();
        PhantomReference<Object> ref = new PhantomReference<Object>(obj, queue);
        obj = null;

        System.gc();

        @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
        List<Object> storage = new ArrayList<Object>();
        try {
            while (queue.poll() == null) {
                storage.add(new byte[1024 * 1024]); // 1M
            }
        } catch (OutOfMemoryError ignored) {
        }
        ref.hashCode();
        storage.clear();

        try {
            TimeUnit.MILLISECONDS.sleep(10);
        } catch (InterruptedException ignored) {
        }
    }
```

Honestly neither solution seems great. The latter is a bit better imo, because it checks that reference processing actually happened, but it's still far from ideal. What do you think?